halo2_proofs/src/plonk/permutation/keygen.rs

```diff
@@ -109,49 +109,58 @@ impl Assembly {
         p: &Argument,
     ) -> VerifyingKey<C> {
         // Compute [omega^0, omega^1, ..., omega^{params.n - 1}]
-        let mut omega_powers = Vec::with_capacity(params.n() as usize);
+        let mut omega_powers = vec![C::Scalar::zero(); params.n() as usize];
         {
-            let mut cur = C::Scalar::one();
-            for _ in 0..params.n() {
-                omega_powers.push(cur);
-                cur *= &domain.get_omega();
-            }
+            let omega = domain.get_omega();
+            parallelize(&mut omega_powers, |o, start| {
+                let mut cur = omega.pow_vartime(&[start as u64]);
+                for v in o.iter_mut() {
+                    *v = cur;
+                    cur *= &omega;
+                }
+            })
         }
 
         // Compute [omega_powers * \delta^0, omega_powers * \delta^1, ..., omega_powers * \delta^m]
-        let mut deltaomega = Vec::with_capacity(p.columns.len());
+        let mut deltaomega = vec![omega_powers; p.columns.len()];
         {
-            let mut cur = C::Scalar::one();
-            for _ in 0..p.columns.len() {
-                let mut omega_powers = omega_powers.clone();
-                for o in &mut omega_powers {
-                    *o *= &cur;
+            parallelize(&mut deltaomega, |o, start| {
+                let mut cur = C::Scalar::DELTA.pow_vartime(&[start as u64]);
+                for omega_powers in o.iter_mut() {
+                    for v in omega_powers {
+                        *v *= &cur;
+                    }
+                    cur *= &C::Scalar::DELTA;
                 }
+            });
+        }
 
-                deltaomega.push(omega_powers);
-
-                cur *= &C::Scalar::DELTA;
-            }
+        // Computes the permutation polynomial based on the permutation
+        // description in the assembly.
+        let mut permutations = vec![domain.empty_lagrange(); p.columns.len()];
+        {
+            parallelize(&mut permutations, |o, start| {
+                for (x, permutation_poly) in o.iter_mut().enumerate() {
+                    let i = start + x;
+                    for (j, p) in permutation_poly.iter_mut().enumerate() {
+                        let (permuted_i, permuted_j) = self.mapping[i][j];
+                        *p = deltaomega[permuted_i][permuted_j];
+                    }
+                }
+            });
```

Comment:
Also can update to parallelize inner loop here.

```suggestion
            for (i, permutation_poly) in permutations.iter_mut().enumerate() {
                parallelize(permutation_poly, |permutation_poly, start| {
                    for (p, j) in permutation_poly.iter_mut().zip(start..) {
                        let (permuted_i, permuted_j) = self.mapping[i][j];
                        *p = deltaomega[permuted_i][permuted_j];
                    }
                });
            }
```